File: bench/concurrency/2.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <chrono>

using namespace std;

int x = 0;
mutex x_mutex;

void subroutine()
{
    x_mutex.lock();
	x++;
    x_mutex.unlock();
}

int main(int argc, char const *argv[])
{
	int n = 25000; // Number of threads
	if (argc == 2)
	{
		n = atoi(argv[1]);
	}

	auto start = chrono::steady_clock::now();

	vector<thread> ths;
	for (int i = 0; i < n; i++)
	{
		ths.push_back(thread(&subroutine));
	}

	for (auto &th : ths)
	{
		th.join();
	}

	cout << chrono::duration_cast<chrono::nanoseconds>(chrono::steady_clock::now() - start).count() << endl;

	return 0;
}

File: bench/concurrency/2.go
package main

import (
	"os"
	"strconv"
	"sync"
	"fmt"
	"time"
)

var x = 0
var mux sync.Mutex

func subroutine() {
	mux.Lock()
	x++
	mux.Unlock()
}

func main() {
	var n = 2000000 // Number of goroutines
	var args = os.Args[1:]
	if len(args) >= 1 {
		n, _ = strconv.Atoi(args[0])
	}

	var start = time.Now()

	for i := 0; i < n; i++ {
		go subroutine()
	}

	fmt.Println(time.Now().Sub(start).Nanoseconds())
}
